### DISCUSSION TITLE
Single hook vs multiple hooks

### DISCUSSION THREAD

**Discussion by paganwinter** [fastify/fastify#3669] [Q&A] [answered]
Suppose I have multiple things that I need to do sequentially in a particular lifecycle hook, which is a more optimal/efficient approach run-time performance-wise?

The only way this impacts me other than performance-wise, is with respect to code maintainability and testablity (unit testing of individual hooks).

**Options 1**
```js
app.addHook('onRequest', hookOne)
app.addHook('onRequest', hookTwo)
app.addHook('onRequest', hookThree)
```

**Options 2**
```js
app.addHook('onRequest', async (req, rep) => {
  await hookOne(req, rep)
  await hookTwo(req, rep)
  await hookThree(req, rep)
})
```

Thanks in advance.

**Comment by mcollina** [comment, marked as answer]
I would do the second option.

**Reply by paganwinter** [reply]
Thanks so much for the super quick reply :D

Is it because of some gain from not having Fastify loop through multiple hook functions to invoke each one?

**Reply by mcollina** [reply]
Exactly

**Reply by paganwinter** [reply]
Thanks once again.
On a somewhat related note, had another query.
I am trying to build a framework that wraps around Fastify, to implement some standard behaviour in developed services.

I have a bunch of individual hooks that all services built with my framework need to run.
I have been trying to find a nice way to implement them so that it's easy to maintain and test the framework itself.
I have the following options; of these what would you suggest is a good option?

Again, it's only for better maintainability, but does that introduce any performance penalty at run-time?

**Option 1**
```js
// app.js
app.addHook('onRequest', onRequestHooks);

// hooks.js
async function onRequestHooks(req, rep) {
  await hookOne(req, rep);
  await hookTwo(req, rep);
  await hookThree(req, rep);
}
```

**Option 2**
```js
// app.js
await setupOnRequestHooks(app);

// hooks.js
async function setupOnRequestHooks(app) {
  app.addHook('onRequest', async (req, rep) => {
    await hookOne(req, rep);
    await hookTwo(req, rep);
    await hookThree(req, rep);
  });
}
```

**Reply by mcollina** [reply]
```js
const fp = require('fastify-plugin')
const plugin = async function (app) {
  app.addHook('onRequest', async (req, rep) => {
    await hookOne(req, rep);
    await hookTwo(req, rep);
    await hookThree(req, rep);
  });
}

app.register(plugin)
```

**Reply by paganwinter** [reply]
Does `fastify-plugin` not add any overhead?

I wanted to avoid it, in favour of having the structure of the entire application be more easily readable in a single place.

**Reply by mcollina** [reply]
It doesn't.
